lib/quicly.c

```diff
@@ -4462,25 +4555,65 @@ static int handle_payload(quicly_conn_t *conn, size_t epoch, const uint8_t *_src
         /*   +----------------------+----+----+----+----+---------------+ */
 #undef FRAME
     };
+    static const struct {
+        uint64_t type;
+        struct st_quicly_frame_handler_t _;
+    } ex_frame_handlers[] = {
+#define FRAME(uc, lc, i, z, h, o, ae)                                                                                              \
+    {                                                                                                                              \
+        QUICLY_FRAME_TYPE_##uc,                                                                                                    \
+        {                                                                                                                          \
+            handle_##lc##_frame,                                                                                                   \
+            (i << QUICLY_EPOCH_INITIAL) | (z << QUICLY_EPOCH_0RTT) | (h << QUICLY_EPOCH_HANDSHAKE) | (o << QUICLY_EPOCH_1RTT),     \
+            ae                                                                                                                     \
+        },                                                                                                                         \
+    }
+        /*   +-------------------------------+-------------------+---------------+
+         *   |             frame             |  permitted epochs |               |
+         *   |---------------+---------------+----+----+----+----+ ack-eliciting |
+         *   |  upper-case   |  lower-case   | IN | 0R | HS | 1R |               |
+         *   +---------------+---------------+----+----+----+----+---------------+ */
+        FRAME( ACK_FREQUENCY , ack_frequency ,  0 ,  0 ,  0 ,  1 ,             1 ),
```

Comment:
What ar ethe upper case and lower case used for again?

Reply:
Yeah we need both, because the frame types are defined using upper-case letters (e.g., QUICLY_FRAME_TYPE_ACK_FREQUENCY), while the callback functions (e.g., handle_ack_frequency_frame) is defined in lower-case.

Unfortunately, there's no way of changing the case using a macro.